### PR TITLE
Handle more commands in Spirc

### PR DESCRIPTION
**Pull request by joerg-krause**
Add a handler for the commands: _Play, PlayPause, Pause, Prev, Next, VolumeUp_, and _VolumeDown_ to Spirc.

### CODE REVIEW

File: src/spirc.rs [hunk 1]
```diff
@@ -167,6 +174,27 @@ impl Spirc {
         (spirc, task)
     }
 
+    pub fn play(&self) {
+        let _ = mpsc::UnboundedSender::send(&self.commands, SpircCommand::Play);
+    }
+    pub fn playpause(&self) {
```
Comment (plietar) [inline src/spirc.rs]: Put an underscore: `play_pause`

File: src/spirc.rs [hunk 2]
```diff
@@ -167,6 +174,27 @@ impl Spirc {
...
+    pub fn pause(&self) {
+        let _ = mpsc::UnboundedSender::send(&self.commands, SpircCommand::Pause);
+    }
+    pub fn prev(&self) {
+        let _ = mpsc::UnboundedSender::send(&self.commands, SpircCommand::Prev);
+    }
+    pub fn next(&self) {
+        let _ = mpsc::UnboundedSender::send(&self.commands, SpircCommand::Next);
+    }
+    pub fn volumeup(&self) {
```
Comment (plietar) [inline src/spirc.rs]: `volume_up`

File: src/spirc.rs [hunk 3]
```diff
@@ -167,6 +174,27 @@ impl Spirc {
...
+        let _ = mpsc::UnboundedSender::send(&self.commands, SpircCommand::VolumeUp);
+    }
+    pub fn volumedown(&self) {
```
Comment (plietar) [inline src/spirc.rs]: `volume_down`

File: src/spirc.rs [hunk 4]
```diff
@@ -363,6 +411,45 @@ impl SpircTask {
                 self.notify(None);
             }
 
+            MessageType::kMessageTypeVolumeUp => {
+                let old_vol = self.mixer.volume();
+
+                // Convert the absolute value into a relative value. Than add 10%, check
+                // for boundaries and convert the new relative volume back to an absolute value.
```
Comment (plietar) [inline src/spirc.rs]: Why does need to be converted to float ? You can just add 6554 as an integer (there's a small rounding error, but that's fine IMO)

File: src/spirc.rs [hunk 5]
```diff
@@ -277,28 +328,25 @@ impl SpircTask {
...
+            MessageType::kMessageTypePlayPause => {
                 if self.state.get_status() == PlayStatus::kPlayStatusPlay {
-                    self.player.pause();
-                    self.mixer.stop();
-                    self.state.set_status(PlayStatus::kPlayStatusPause);
-
-                    let now = now_ms() as u64;
-                    let position = self.state.get_position_ms();
+                    self.handle_pause();
+                } else {
```
Comment (plietar) [inline src/spirc.rs]: That should be `else if self.state.get_status() == PlayStatus::kPlayStatusPause`

File: src/spirc.rs [hunk 6]
```diff
@@ -378,6 +465,27 @@ impl SpircTask {
         }
     }
 
+    fn handle_play(&mut self) {
```
Comment (plietar) [inline src/spirc.rs:434]: Add `assert_eq!(self.state.get_status(), PlayStatus::kPlayStatusPause)` and document it as a precondition

File: src/spirc.rs [hunk 7]
```diff
@@ -378,6 +465,27 @@ impl SpircTask {
...
+        self.mixer.start();
+        self.player.play();
+        self.state.set_status(PlayStatus::kPlayStatusPlay);
+        self.state.set_position_measured_at(now_ms() as u64);
+    }
+
+    fn handle_pause(&mut self) {
```
Comment (plietar) [inline src/spirc.rs]: Add `assert_eq!(self.state.get_status(), PlayStatus::kPlayStatusPause)`

File: src/spirc.rs [hunk 8]
```diff
@@ -246,7 +295,9 @@ impl SpircTask {
 
         if frame.get_ident() == self.ident ||
            (frame.get_recipient().len() > 0 && !frame.get_recipient().contains(&self.ident)) {
-            return;
+            if frame.get_typ() == MessageType::kMessageTypeNotify {
```
Comment (plietar) [inline src/spirc.rs]: Not too comfortable with that, it'll start reacting to messages targeted at other devices.

I don't think we should send messages to ourselves. I suggest making `handle_command` do `if active { self.handle_play(); } else { broadcast_message(kMessageTypePlay); }`

File: src/spirc.rs [hunk 9]
```diff
@@ -228,6 +256,27 @@ impl Future for SpircTask {
 impl SpircTask {
     fn handle_command(&mut self, cmd: SpircCommand) {
         match cmd {
+            SpircCommand::Play => {
+                CommandSender::new(self, MessageType::kMessageTypePlay).send();
```
Comment (plietar) [inline src/spirc.rs]: Broadcasting these messages isn't great, ideally we'd want to track the currently active device.
It's probably fine for now, we'll assume clients just ignore them if they aren't active.

Can you please add a TODO though.

**Review by plietar — CHANGES_REQUESTED** [review]
Thanks, looks good overall.

Could you also add volume and load ?
```rust
impl Spirc {
  fn volume(&self, volume: u16) { ... }

  // device is the ident, not the name.
  // If device is None load on self
  fn load(device: Option<String>, tracks: Vec<SpotifyId>, track_index: usize, play: bool, position: u32) { ... }
}
```

For `load` we will want a way of discovering devices as well, but that can be added later

**Comment by plietar** [comment]
While you're at it, could you rename `CommandSender` to `MessageSender`, since command now refers to the `Spirc -> SpircTask`, and `handle_frame` to `handle_message`.

**Comment by plietar** [comment]
I've never seen the `VolumeUp`/`VolumeDown`/`PlayPause` messages, which is why I've not bothered handling them.

Do you have a device which sends these ? I'm not objecting to implementing them, just curious.

**Comment by joerg-krause** [comment]
> I've never seen the VolumeUp/VolumeDown/PlayPause messages, which is why I've not bothered handling them. Do you have a device which sends these ? I'm not objecting to implementing them, just curious.

I am using a device with some multimedia keys attached to GPIOs. So, my device is generating those commands, and the Spotify player gladly handles them.

**Comment by plietar** [comment]
> I am using a device with some multimedia keys attached to GPIOs. So, my device is generating those commands, and the Spotify player gladly handles them.

So this is your code sending them right ? Seems like Spotify implemented them on the receiving side but none of their implementations actually send them.

File: src/spirc.rs (hunk 8 above)
Comment (joerg-krause) [reply src/spirc.rs]: I'd like to use librespot with an Unix socket listener listening for commands generated by another process running on the same board. The other process handles input events like _PlayPause_. I don't want to have to instances of a librespot running, one for streaming and one for remote control, but only one librespot instance being capable of doing both. Note, that the size of a librespot binary is more than 2.1 MB and my device is limited in both flash and RAM memory.

Comment (plietar) [reply src/spirc.rs]: I'm not arguing librespot shouldn't be able to control itself, just that there's no need to broadcast a message out to do it.
If librespot is currently active we can just process the command locally.

This is how Spotify's implementations do. When playing locally they don't broadcast messages about what they're doing (well they do send `kMessageTypeNotify`, and so should librespot)

**Comment by joerg-krause** [comment]
Yes, I am using a Unix socket listener with librespot: https://github.com/joerg-krause/librespot/commit/3a03d21348e212235fc48a13cb3a5e30f356562a. The messages are generated by the running librespot instance.

File: src/spirc.rs (hunk 8 above)
Comment (joerg-krause) [reply src/spirc.rs]: I see! I wasn't aware that these messages are broadcasted.

File: src/spirc.rs (hunk 4 above)
Comment (joerg-krause) [reply src/spirc.rs]: You're right! No need for float here...

File: src/spirc.rs (hunk 8 above)
Comment (joerg-krause) [reply src/spirc.rs]: @plietar I've tested with the following code snippet:

``` Rust
fn handle_command(&mut self, cmd: SpircCommand) {
    SpircCommand::Pause => {
        if self.device.get_is_active() {
            self.handle_pause();
        } else {
             CommandSender::new(self, MessageType::kMessageTypePause).send();
        }
    }
}
```

librespot stops playback, but the Spotify Player is not notified about it. The Play/Pause button is still showing the pause icon and the progress bar for the seconds is still increasing.

So, these messages needs to be broadcasted so the Spotify Player is notified about the changes, right?

Comment (plietar) [reply src/spirc.rs]: You only need to broadcast the `kMessageTypeNotify` using `self.notify(None);`, not the `kMessageTypePause`.

Comment (joerg-krause) [reply src/spirc.rs]: Thanks! This works:

``` Rust
fn handle_command(&mut self, cmd: SpircCommand) {
        SpircCommand::Pause => {
        if self.device.get_is_active() {
            self.handle_pause();
        }
        self.notify(None);
    }
}
```

Should `self.notify(None);` better be placed inside the `if` branch`?

**Comment by joerg-krause** [comment]
@plietar I've update the commit. I did not added `volume()` and `load()` yet, as I do not know how to pass the variables, e.g. `volume`, to `SpircTask`. But this can be done in a follow-up patch...

**Comment by mattiash** [comment]
> Seems like Spotify implemented them on the receiving side but none of their implementations actually send them.

I have a Raumfeld One S which has a built-in spotify connect client and physical volume up/down buttons. If I play music with spotify connect and press those buttons, I can see the volume change in spotify. So it seems that the Raumfeld tells the spotify server that it has changed the volume.

I am very interested in this PR since I want to build something with similar functionality as the Raumfeld One S with a Raspberry PI, an external amplifier and a remote that allows me to change the volume and play/pause the music.

File: src/spirc.rs [hunk 10]
```diff
@@ -227,12 +255,42 @@ impl Future for SpircTask {
 
 impl SpircTask {
     fn handle_command(&mut self, cmd: SpircCommand) {
+        let active = self.device.get_is_active();
         match cmd {
+            SpircCommand::Play if active => {
```
Comment (plietar) [inline src/spirc.rs]: This isn't handling the `!active` case anymore.
That code should look like 
```rust
match cmd {
    SpircCommand::Play => {
        if active {
            self.handle_play();
            self.notify(None);
        } else {
            CommandSender::new(self, MessageType::kMessageTypePlay).send();
        }
    }
```

Comment (joerg-krause) [reply src/spirc.rs]: For all commands (except Shutdown)?

Comment (plietar) [reply src/spirc.rs]: Yes, It's quite a lot of boilerplate unfortunately

**Comment by joerg-krause** [comment]
@plietar I've updated the PR.

EDIT: Looks like the rustup script failed.

**Comment by plietar** [comment]
Thanks to @herrernst travis is working again.
@joerg-krause could you rebase to get the fix please ?

**Comment by joerg-krause** [comment]
@plietar Did the rebase, but travis is not running anymore?

**Review by plietar — APPROVED** [review]

**Comment by plietar** [comment]
Thanks !

**Comment by cortegedusage** [comment]
this might be a very ignorant question, 
but does this mean that librespot now can be controlled from the command line with someting simple as librespot play/pause/next?
or does this still involve some basic programming skills.

I'd like to control a playlist with a remote control that's why I'm asking.

**Comment by plietar** [comment]
This is only an API applications which depend on librespot can use.
@joerg-krause exposes it as a Unix socket in his fork https://github.com/joerg-krause/librespot/commit/bc457711b54e0b6cde598e05f7876f2a528c807d

**Comment by cortegedusage** [comment]
Thank you for your quick reply,
I'm afraid this is something way out of my league.
But as long as the music is playing, I can only bow my head in respect for
people who can code. 

keep up the good work!

**Comment by seanrand** [comment]
@plietar would you consider making it possible to expose Spirc as a Unix socket, or do you think it's out of scope for librespot? I noticed that lots of refactoring happened between @joerg-krause's patch and current master, so my idea of simply hacking it into my local RPi instance of librespot is not as straight forward as i thought. Some sort of external API, however unstable it might be is the only thing I'm missing right now.

**Comment by joerg-krause** [comment]
@seanrand Please have a look at my current https://github.com/joerg-krause/librespot/tree/tokio-uds branch, which is even with @plietar master branch, and adds a Unix domain socket.

I guess it is more a feature for https://github.com/Spotifyd/spotifyd.

**Pull request merged**